signature/internal/base/envelope.go

```diff
@@ -0,0 +1,213 @@
+	err := validateSignRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := e.Envelope.Sign(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate certificate chain
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+	signerInfo := content.SignerInfo
+
+	if err := validateCertificateChain(
+		signerInfo.CertificateChain,
+		signerInfo.SignedAttributes.SigningTime,
+		signerInfo.SignatureAlgorithm,
+	); err != nil {
+		return nil, err
+	}
+
+	e.Raw = raw
+	return e.Raw, nil
+}
+
+// Verify performs integrity and other signature specification related
+// validations.
+// It returns envelope content containing the payload to be signed and
+// SignerInfo object containing the information about the signature.
+//
+// Reference: https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#steps
+func (e *Envelope) Verify() (*signature.EnvelopeContent, error) {
+	// validation before the core verify process.
+	if len(e.Raw) == 0 {
+		return nil, &signature.MalformedSignatureError{}
+	}
+
+	// core verify process.
+	content, err := e.Envelope.Verify()
+	if err != nil {
+		return nil, err
+	}
+
+	// validation after the core verify process.
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// Content returns the validated signature information and payload.
+func (e *Envelope) Content() (*signature.EnvelopeContent, error) {
+	if len(e.Raw) == 0 {
+		return nil, &signature.MalformedSignatureError{Msg: "raw signature is empty"}
+	}
+
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// validateSignRequest performs basic set of validations on SignRequest struct.
+func validateSignRequest(req *signature.SignRequest) error {
+	if err := validatePayload(&req.Payload); err != nil {
+		return err
+	}
+
+	if err := validateSigningTime(req.SigningTime, req.Expiry); err != nil {
+		return err
+	}
+
+	if req.Signer == nil {
+		return &signature.MalformedSignatureError{Msg: "signer is nil"}
+	}
+
+	_, err := req.Signer.KeySpec()
+	return err
+}
+
+// validateEnvelopeContent validates the content which includes signerInfo and
+// payload.
+func validateEnvelopeContent(content *signature.EnvelopeContent) error {
+	if err := validatePayload(&content.Payload); err != nil {
+		return err
+	}
+	return validateSignerInfo(&content.SignerInfo)
+}
+
+// validateSignerInfo performs basic set of validations on SignerInfo struct.
+func validateSignerInfo(info *signature.SignerInfo) error {
+	if len(info.Signature) == 0 {
+		return &signature.MalformedSignatureError{Msg: "signature not present or is empty"}
+	}
+
+	if info.SignatureAlgorithm == 0 {
+		return &signature.MalformedSignatureError{Msg: "SignatureAlgorithm is not present"}
+	}
+
+	signingTime := info.SignedAttributes.SigningTime
```

Comment:
We are missing validation for SigningScheme, SigningAgent.

Reply:
@priteshbandi Can you point out the implementation of the validation before this PR so that they can be included.

Reply:
- [SigningScheme](https://github.com/notaryproject/notation-core-go/blob/main/signer/signer.go#L232-L234)
- Looks like there is no validation for SigningAgent